### ISSUE TITLE
Broker Latency

### ISSUE THREAD

**Issue by mongldroid** [martin-ger/uMQTTBroker#31]
Currently running the broker on an esp8266. The idea is to have a plethera of client devices where there is one device on each drum in my kit which is getting readings from a piezo vibration sensor.

The broker and everything seems to be running fine technically speaking however I have noticed when I have even just a short burst of publishes (issues even with three) they broker is seemingly slow to process them.

For example if I play quarter notes there is 0 noticeable latency and the broker recieves the velocity of the drum hit with near real time. Yet if I play a 16th bar. the messages lag into the server over a broader time. (If I hit 5 times in a second, it may take 2 seonds to play all the notes on the broker side, but when hitting one note every second there is 0 latency)

The debounce on the trigger is 80ms so in testing with jsut one client no messages are coming faster than 80ms and there is still almost a second of lag when a series of quick hits comes in.

Is there a way to optimize this. I cannot for the life of me understand why the lag happens considering this is basically a c++ FIFO buffer server similar to class projects I have done. This should be working at microsecond level, let alone 80ms cycle lvl, correct?

The LED on the transmitter is set to flash on each publish and operates perfectly even with rapid drumming. no lag, perfect timing.

I am not using any delays.

Should it lag this much just on 4 messages over 200ms?

[esp_broker.txt](https://github.com/martin-ger/uMQTTBroker/files/2907266/esp_broker.txt)
[esp_client.txt](https://github.com/martin-ger/uMQTTBroker/files/2907267/esp_client.txt)

**Comment by martin-ger** [comment]
Interesting application! Actually, I havn't done any performance measurements right now, but I would also like to understand it...

What happens, if you do just a println("x"); instead of the Midi-call on the broker - same delay?

**Comment by mongldroid** [comment]
correct, that is actually the primray metric i have used.. watching the
serial monitor to compare my hits to look for visual lag. i thought maybe
the print may be slowing things down but should be no different than
sending the midi cmd

i just cannot understand why latency only happens on burst but works
flawless.

**Comment by martin-ger** [comment]
We are talking about TCP here - this means we acks going back. Maybe they are delayed. Could you do the same test using a Mosquitto broker on a Linux/Raspi box? Just to make sure, that the delay really happens on the broker?

**Comment by mongldroid** [comment]
I can try that. I am quit confident the lag is not in the broker itself
because, cmon its a FIFO buffer in c++, it doesn't really get faster or
more efficient hardly.

I am wondering if you can think of anything thta might help the lag, which
is probably from TCP. Maybe parraelize it a bit with more than one TCP
connection or something. Maybe you can suggest an alternate way of sending
exactly 4byte packets.

Ive tried UDP but reliability made it a nonsolution (every hit has to make
it in music lol), I thought for sure this was the solution. Maybe a simple
moded web server would work faster.

Maybe I need to make my "mothership" a raspberry pi with a full mosquito,
maybe an esp just wont cut the mustard for what I need, but I cannot
imagine why it could not be done.

**Comment by RudyFiero** [comment]
As far as UDP. Send three times with a small delay between. On the receive end, if the message is the same as the last, then discard. You need a message counter just to identify if it is a duplicate.

**Comment by mongldroid** [comment]
The latency is agreeably in TCP as you suggest. I wrote up some better UDP and things move without lag. 

So from this, it brings up a question for me. Since MQTT has multilevel quality of service, why not implement qos as udp, it fits the definition pretty nicely and then do TCP handshake for qos >0
